src/waveform/renderers/allshader/digitsrenderer.cpp

```diff
@@ -0,0 +1,176 @@
+    QFont font;
+    const char* str = "0123456789:.";
+    font.setFamily("Open Sans");
+    font.setPixelSize(18);
+
+    QFontMetricsF metrics{font};
+
+    qreal totalTextWidth = 0;
+    qreal maxTextHeight = 0;
+    for (int i = 0; i < 12; i++) {
+        assert(charToIndex(str[i]) == i);
+        const QString text(str[i]);
+        const auto rect = metrics.tightBoundingRect(text);
+        maxTextHeight = std::max(maxTextHeight, rect.height());
+        totalTextWidth += metrics.horizontalAdvance(text) + 6;
+    }
+
+    totalTextWidth = std::ceil(totalTextWidth);
+
+    QImage image(static_cast<int>(totalTextWidth * devicePixelRatio),
+            static_cast<int>(std::ceil(maxTextHeight + 6) * devicePixelRatio),
+            QImage::Format_ARGB32_Premultiplied);
+    image.setDevicePixelRatio(devicePixelRatio);
+    image.fill(Qt::transparent);
+
+    // Draw digits with dark outline
+    QPainter painter;
+    QPen pen(QColor(0, 0, 0, 192));
+    pen.setWidth(3);
+
+    painter.begin(&image);
+    painter.setRenderHint(QPainter::Antialiasing);
+    painter.setWorldMatrixEnabled(false);
+    painter.setBrush(QColor(0, 0, 0, 224));
+    painter.setPen(pen);
+    painter.setFont(font);
+    qreal x = 0;
+    for (int i = 0; i < 12; i++) {
+        const QString text(str[i]);
+        QPainterPath path;
+        path.addText(QPointF(x + 3, maxTextHeight + 3), font, text);
+        painter.drawPath(path);
+        x += metrics.horizontalAdvance(text) + 6;
+    }
+    painter.end();
+
+    // Apply Gaussian blur to dark outline
+
+    QGraphicsBlurEffect* blur = new QGraphicsBlurEffect; // ownership passed to item
+    blur->setBlurRadius(3);
+
+    QGraphicsScene scene;
+    QGraphicsPixmapItem item;
+    item.setPixmap(QPixmap::fromImage(image));
+    item.setGraphicsEffect(blur);
+    scene.addItem(&item);
+
+    image.fill(Qt::transparent);
+    painter.begin(&image);
+    scene.render(&painter, QRectF(), QRectF(0, 0, image.width(), image.height()));
+
+    // Draw digits foreground
+    painter.setPen(Qt::white);
+    painter.setFont(font);
+
+    x = 0;
+    for (int i = 0; i < 12; i++) {
+        const QString text(str[i]);
+        painter.drawText(QPointF(x + 3, maxTextHeight + 3), text);
+        m_offset[i] = static_cast<float>(x / totalTextWidth);
+        m_width[i] = static_cast<float>(metrics.horizontalAdvance(text) + 6);
+        x += metrics.horizontalAdvance(text) + 6;
+    }
+    m_offset[12] = 1.f;
+
+    painter.end();
+
+    m_pTexture = createTexture(image);
+}
+
+void allshader::DigitsRenderer::draw(const QMatrix4x4& matrix,
+        float x,
+        float y,
+        const QString& s,
+        QColor color,
+        float devicePixelRatio) {
+    const int n = s.length();
+
+    VertexData posVertices;
+    VertexData texVertices;
+
+    posVertices.reserve(n * 6); // two triangles per character
+    texVertices.reserve(n * 6);
+
+    for (int i = 0; i < n; i++) {
+        int index = charToIndex(s[i].cell());
```

Comment:
This code is hard to understand. Maybe you should pass a char array or a QByteArray instead?

Reply:
improved.

Reply:
Can I resolve this, @daschuer ?